File: src/database/jsonDb.js
import CryptoJS from 'crypto-js';
import fs from 'fs-extra';
import path from 'path';
import { jsonParseWithCircularRefs, jsonStringifyWithCircularRefs, writefile } from 'sbg-utility';

/**
 * Simple JSON file database with circular reference support.
 */
class JsonDB {
  /**
   * Directory where JSON files are stored.
   * @type {string}
   */
  directory;

  /**
   * Create a new JsonDB instance.
   * @param {string} directory - Directory to store JSON files.
   */
  constructor(directory) {
    this.directory = directory;
  }

  /**
   * Save data to a JSON file with the given id.
   * @param {string} id - Identifier for the JSON file (without extension).
   * @param {any} data - Data to save (can include circular references).
   * @returns {void}
   */
  save(id, data) {
    id = this.#hash(id);
    const savePath = path.join(this.directory, `${id}.json`);
    const encoded = jsonStringifyWithCircularRefs(data);
    if (typeof encoded === 'string') writefile(savePath, encoded);
  }

  /**
   * Load data from a JSON file with the given id.
   * @template T
   * @param {string} id - Identifier for the JSON file (without extension).
   * @returns {T} The parsed data (with circular references restored).
   */
  load(id) {
    id = this.#hash(id);
    const loadPath = path.join(this.directory, `${id}.json`);
    const parse = jsonParseWithCircularRefs(fs.readFileSync(loadPath, 'utf8'));
    return parse;
  }

  /**
   * Delete a JSON file with the given id.
   * @param {string} id - Identifier for the JSON file (without extension).
   * @returns {void}
   */
  delete(id) {
    id = this.#hash(id);
    const deletePath = path.join(this.directory, `${id}.json`);
    if (fs.existsSync(deletePath)) {
      fs.rmSync(deletePath);
    }
  }

  /**
   * Generate an MD5 hash for the given id.
   * @param {string} id - The input string to hash.
   * @returns {string} The MD5 hash of the id.
   */
  #hash(id) {
    return CryptoJS.MD5(id).toString();
  }
}

export default JsonDB;

File: src/database/jsonDb.py
import os
import hashlib
import jsonpickle


class JsonDB:
    """
    Simple JSON file database with circular reference support.
    """

    def __init__(self, directory: str):
        """
        Create a new JsonDB instance.
        :param directory: Directory to store JSON files.
        """
        self.directory = directory
        os.makedirs(self.directory, exist_ok=True)

    def _hash(self, id: str) -> str:
        """
        Generate an MD5 hash for the given id.
        :param id: The input string to hash.
        :return: The MD5 hash of the id.
        """
        return hashlib.md5(id.encode("utf-8")).hexdigest()

    def save(self, id: str, data):
        """
        Save data to a JSON file with the given id.
        :param id: Identifier for the JSON file (without extension).
        :param data: Data to save (can include circular references).
        """
        hashed_id = self._hash(id)
        save_path = os.path.join(self.directory, f"{hashed_id}.json")
        encoded = jsonpickle.encode(data, make_refs=True)
        if isinstance(encoded, str):
            with open(save_path, "w", encoding="utf-8") as f:
                f.write(encoded)

    def load(self, id: str):
        """
        Load data from a JSON file with the given id.
        :param id: Identifier for the JSON file (without extension).
        :return: The parsed data (with circular references restored).
        """
        hashed_id = self._hash(id)
        load_path = os.path.join(self.directory, f"{hashed_id}.json")
        if not os.path.exists(load_path):
            raise FileNotFoundError(f"No JSON file found for id '{id}'")
        with open(load_path, "r", encoding="utf-8") as f:
            return jsonpickle.decode(f.read())

    def delete(self, id: str):
        """
        Delete a JSON file with the given id.
        :param id: Identifier for the JSON file (without extension).
        """
        hashed_id = self._hash(id)
        delete_path = os.path.join(self.directory, f"{hashed_id}.json")
        if os.path.exists(delete_path):
            os.remove(delete_path)
